### PR TITLE
feat(interpreters): add ExecSandboxInterpreter for hardware-isolated code execution

### LINKED ISSUE #3884: Add exec-sandbox interpreter (self-hosted QEMU microVM sandbox)
# Add exec-sandbox interpreter (self-hosted QEMU microVM sandbox)

## Summary

Proposing [exec-sandbox](https://github.com/dualeai/exec-sandbox) as a new `BaseInterpreter` backend for `CodeExecutionToolkit`. It runs code in ephemeral QEMU microVMs with hardware-level isolation (KVM on Linux, HVF on macOS) — self-hosted, no cloud account needed.

## Why

- **Self-hosted** — E2BInterpreter requires an E2B account, DockerInterpreter requires Docker. exec-sandbox runs on bare metal with just QEMU. Data never leaves the machine — relevant for customer data privacy, regulated environments, and cheaper at scale.
- **macOS + Linux** — HVF on macOS, KVM on Linux. Same codebase for dev and prod. No Docker dependency.
- **Fast** — ~1-2ms from warm pool, ~200ms from memory snapshots.
- **VM isolation** — Hardware-level boundary, same category as MicrosandboxInterpreter but QEMU-based.
- **Apache-2.0**

## How it maps

exec-sandbox is a Python library with an async API. Following the MicrosandboxInterpreter pattern (PR #3069):

| BaseInterpreter method | exec-sandbox mapping |
|---|---|
| `run(code, code_type)` | `session.exec(code)` via language-mapped session |
| `supported_code_types()` | `["python", "py", "py3", "javascript", "js", "bash", "sh"]` |
| `update_action_space(action_space)` | No-op (sandboxed VM, like MicrosandboxInterpreter) |
| `execute_command(command)` | `session.exec(command)` via `language="raw"` session |

```python
import asyncio
from exec_sandbox import Scheduler

class ExecSandboxInterpreter(BaseInterpreter):
    _CODE_TYPE_MAPPING = {
        "python": "python", "py": "python", "py3": "python",
        "javascript": "javascript", "js": "javascript",
        "bash": "raw", "sh": "raw", "shell": "raw",
    }

    def __init__(self, require_confirm: bool = True, timeout: int = 30):
        self._require_confirm = require_confirm
        self._timeout = timeout
        self._loop = asyncio.new_event_loop()
        self._scheduler = self._loop.run_until_complete(Scheduler().__aenter__())
        self._sessions: dict[str, Session] = {}  # language -> session

    def _get_or_create_session(self, language: str):
        if language not in self._sessions:
            self._sessions[language] = self._loop.run_until_complete(
                self._scheduler.session(language=language)
            )
        return self._sessions[language]

    def run(self, code: str, code_type: str) -> str:
        lang = self._CODE_TYPE_MAPPING.get(code_type.lower())
        if not lang:
            raise InterpreterError(f"Unsupported code type: {code_type}")
        session = self._get_or_create_session(lang)
        result = self._loop.run_until_complete(session.exec(code))
        parts = []
        if result.stdout.strip():
            parts.append(result.stdout.strip())
        if result.stderr.strip():
            parts.append(f"STDERR: {result.stderr.strip()}")
        return "\n".join(parts) or "Code executed successfully (no output)"

    def supported_code_types(self) -> list[str]:
        return list(self._CODE_TYPE_MAPPING.keys())

    def update_action_space(self, action_space):
        pass  # no-op: sandboxed VM

    def execute_command(self, command: str) -> str:
        session = self._get_or_create_session("raw")
        result = self._loop.run_until_complete(session.exec(command))
        return result.stdout
```

Usage:

```python
toolkit = CodeExecutionToolkit(sandbox="exec_sandbox")
# or directly:
interpreter = ExecSandboxInterpreter(require_confirm=False)
```

## Open questions

- **Async/sync bridge** — exec-sandbox is fully async. The sketch uses a dedicated event loop. A background-thread loop would be safer if called from an existing async context.
- **Config passthrough** — Following the `microsandbox_config` pattern, an `exec_sandbox_config` dict in `CodeExecutionToolkit` for VM options (memory, timeout, packages).

## Links

- [Repo](https://github.com/dualeai/exec-sandbox) · [PyPI](https://pypi.org/project/exec-sandbox/)
- Reference PR: #3069 (MicrosandboxInterpreter — same integration pattern)

Happy to submit a PR if there's interest.

### PR DESCRIPTION
**Pull request by clemlesne**
## Summary

Closes #3884.

Add [exec-sandbox](https://github.com/duale-ai/exec-sandbox) as a new `BaseInterpreter` backend. Runs code in ephemeral QEMU microVMs with hardware-level isolation (KVM/HVF), self-hosted, no cloud account needed.

- **New `ExecSandboxInterpreter`** with lazy scheduler/session lifecycle — sessions are persistent per language (Python, JavaScript, shell), so variables, imports, and functions survive across `run()` calls
- **`ExecSandboxConfig`** frozen dataclass for typed `SchedulerConfig` passthrough (warm pool, memory, timeouts, S3, package validation, etc.)
- **`CodeExecutionToolkit`** integration via `sandbox="exec_sandbox"`
- **29 real integration tests** — no mocks, runs actual code in QEMU microVMs. Covers normal cases, session persistence, edge cases, weird cases, and error cases
- **Isolated optional dependency group** `camel-ai[exec-sandbox]` to avoid psutil/aiohttp conflicts with agentops and microsandbox (uv `conflicts` declaration)
- **Documentation** in `docs/key_modules/interpreters.md` with code examples and best practices

## ⚠️ Breaking change: new dependency groups

`agentops` and `microsandbox` have been moved out of `camel-ai[all]` into their own optional dependency groups. This is required because they have conflicting transitive dependencies (`agentops` pins `psutil<7.0.1`, `microsandbox` pins `aiohttp<3.11`) that prevent adding `exec-sandbox` to the resolver.

**Before:**
```bash
pip install "camel-ai[all]"  # included agentops and microsandbox
```

**After:**
```bash
pip install "camel-ai[all]"              # no longer includes agentops or microsandbox
pip install "camel-ai[all,agentops]"     # to get agentops
pip install "camel-ai[all,microsandbox]" # to get microsandbox
pip install "camel-ai[exec-sandbox]"     # new, requires Python 3.12+
```

These groups are declared as conflicting in `pyproject.toml` via uv `conflicts`, so the resolver will not try to install incompatible combinations:

```toml
[tool.uv]
conflicts = [
    [{ extra = "agentops" }, { extra = "exec-sandbox" }],
    [{ extra = "microsandbox" }, { extra = "exec-sandbox" }],
]
```

### Full dependency changes

| Group | Before | After |
|-------|--------|-------|
| `exec-sandbox` | — | New: `exec-sandbox~=0.18; python_version >= '3.12'` |
| `agentops` | In `dev` and `all` | Own group (conflicts with exec-sandbox via psutil) |
| `microsandbox` | In `dev_tools` and `all` | Own group (conflicts with exec-sandbox via aiohttp) |
| `psutil` | `>=5.9.8,<6` | `>=5.9.8,<8` |

### Files changed

| File | Change |
|------|--------|
| `camel/interpreters/exec_sandbox_interpreter.py` | New interpreter + config dataclass |
| `camel/interpreters/__init__.py` | Export new classes |
| `camel/toolkits/code_execution.py` | Add `exec_sandbox` sandbox type |
| `test/interpreters/test_exec_sandbox_interpreter.py` | 29 integration tests |
| `docs/key_modules/interpreters.md` | Add exec-sandbox cards + best practices |
| `pyproject.toml` | Dependency groups, conflicts, psutil relaxation |
| `docs/mintlify/cookbooks/**` | Fix agentops install instructions (`[all]` → `[all,agentops]`) |

## Test plan

- [x] 29/29 integration tests pass (`pytest test/interpreters/test_exec_sandbox_interpreter.py -v` — 25s)
- [x] `ruff check` and `ruff format` clean
- [ ] CI passes (tests skip gracefully when exec-sandbox is not installed)
- [ ] Verify `pip install 'camel-ai[exec-sandbox]'` resolves correctly
- [ ] Verify `pip install 'camel-ai[all]'` still works (exec-sandbox excluded)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by clemlesne** [comment]
> can you help add an example code

I didn't understood what do you mean?

**Comment by clemlesne** [comment]
@fengju0213 Done! Added an example file in 2319b495:

**`examples/interpreters/exec_sandbox_interpreter_example.py`**

It covers:
1. Basic Python execution
2. Session persistence — variables and functions survive across `run()` calls (uses `calculate` from step 1 in step 2)
3. JavaScript execution (separate REPL)
4. Shell commands
5. `CodeExecutionToolkit` integration via `sandbox="exec_sandbox"`

A single `ExecSandboxInterpreter` instance is shared across all examples to demonstrate the scheduler's VM reuse.

**Comment by fengju0213** [comment]
> @fengju0213 Done! Added an example file in [2319b49](https://github.com/camel-ai/camel/commit/2319b495ed7a2a121873ff80eb5f6525fead8f1c):
> 
> **`examples/interpreters/exec_sandbox_interpreter_example.py`**
> 
> It covers:
> 
> 1. Basic Python execution
> 2. Session persistence — variables and functions survive across `run()` calls (uses `calculate` from step 1 in step 2)
> 3. JavaScript execution (separate REPL)
> 4. Shell commands
> 5. `CodeExecutionToolkit` integration via `sandbox="exec_sandbox"`
> 
> A single `ExecSandboxInterpreter` instance is shared across all examples to demonstrate the scheduler's VM reuse.

thanks! @clemlesne

File: camel/interpreters/exec_sandbox_interpreter.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+
+if TYPE_CHECKING:
+    from exec_sandbox import Scheduler, Session
+    from exec_sandbox.models import Language
+
+logger = get_logger(__name__)
+
+
+@dataclass(frozen=True)
+class ExecSandboxConfig:
```
Comment (fengju0213) [inline camel/interpreters/exec_sandbox_interpreter.py:32]: For the sake of stylistic consistency, it would be better if we put this parameter in the interpreter's __init__.
